### ISSUE TITLE
`pnpm` does not update `node_modules` when `node-linker=hoisted`

### ISSUE THREAD

**Issue by cusxio** [pnpm/pnpm#4642]
It seems that `pnpm` always run `--lockfile-only=true` when `node-linker=hoisted`.

https://github.com/pnpm/pnpm/blob/c38feff08f9a110ebab918c6a9a64ad38613adb3/packages/core/src/install/index.ts#L1038-L1042

### Code to reproduce the issue:

1. Create a `.npmrc`, add `node-linker=hoisted` in any existing project.
2.  `pnpm i` with any existing `package.json` to build the `node_modules` folder.
3. Add any new package, i.e. `pnpm add nx`

### Expected behavior:

`pnpm add` / `pnpm up` should update the `node_modules` folder

### Actual behavior:

The following warning popups when `pnpm add` or `pnpm up` is run

```bash
 WARN  `node_modules` is present. Lockfile only installation will make it out-of-date
```

**Comment by zkochan** [comment]
It seems to work fine. The warning should not be printed.

**Comment by cusxio** [comment]
@zkochan truthfully I did not check the `node_modules` folders exactly. I took the warning at face value.

However, in my monorepo setup, after running for example `pnpm up -iLr`, it definitely breaks `eslint`, where `eslint` reports plugin configuration conflicts.

I have to do something like `rm -rf node_modules **/node_modules** pnpm-lock.yaml` to fix the `eslint` conflict issue after upgrading.

**Comment by zkochan** [comment]
OK, if you can create exact steps to reproduce, we can check.

**Comment by sannajammeh** [comment]
Yeah I am still getting this issue. It runs the full installation when node-linker is hoisted

**Comment by drewxs** [comment]
I'm getting this issue as well.

**Comment by nicognaW** [comment]
<img width="914" alt="image" src="https://github.com/pnpm/pnpm/assets/66731869/30b19e8f-87c7-4a95-9ba8-05150e0c25ee">
Just got the issue with similar situation

**Comment by sannajammeh** [comment]
Incredibly annoying as PNPM ends up doing a full complete reinstall of all packages whenever you run install. Any install no matter how small the package will prompt 1 minute install time because pnpm runs postinstall of 12 native deps.

**Comment by rtmann** [comment]
Still happening, exact same issue.  Sort of makes pnpm a non starter for us.

**Comment by dsvgl** [comment]
here too. We always have to completely wipe out all node_modules in our monorepo before updating dependencies.

**Comment by sunxyw** [comment]
After 2 years this still exists, any fix?

**Comment by magnusriga** [comment]
Same issue here. It started happening after setting `node-linker` to `hoisted`. Any way to avoid a complete reinstall of all paclages when only one package is added? Why exactly does it do that?

**Comment by ProdanSergey** [comment]
I have the same issue. As my stack demands a specific testing framework, I had to switch to the `hoisted` strategy. It became reproducible exactly after I had done it.

**Comment by zkochan** [comment]
I don't get it how the comments are related to this issue. The issue is about a warning message which should not be printed.

> It seems that pnpm always run --lockfile-only=true when node-linker=hoisted.

This is the correct behaviour for `node-linker=hoisted`. pnpm updates the lockfile then it hoists the packages using info from the lockfiles. Lastly, it checks what is in the filesystem inside node_modules and makes changes if they are required.

If you have issues with `node-linker=hoisted`, stope commenting in this issue. Create a new issue with steps to reproduce.

**Comment by vladshcherbin** [comment]
how to solve this? trying to use workspaces and I have no clue why on earth this warning pops up upon a simple pnpm add 😵‍💫🔫

**Comment by jgarplind** [comment]
If I understand @zkochan correctly, the warning is moot and should not be printed. That is a (small) bug which this issue is about.

For now we have to live with the warning, until someone contributes a small reproduction which can be used as reference when producing a fix.

**Comment by BabyLy233** [comment]
> how to solve this? trying to use workspaces and I have no clue why on earth this warning pops up upon a simple pnpm add 😵‍💫

I have the same problem. It's my first time to use default setting to use monorepo, but when use `pnpm add` in workspace this warning come up.
Don't know how to fix it or just ignore this.

**Comment by phuctm97** [comment]
I have this same problem with both pnpm 9 and 10 for just a simple monorepo setup. I have no idea what I did wrong.

**Comment by phuctm97** [comment]
Adding this to `.npmrc` fixed the issue for me:

```
shared-workspace-lockfile=true
```

**Comment by BabyLy233** [comment]
> Adding this to `.npmrc` fixed the issue for me:
> 
> ```
> shared-workspace-lockfile=true
> ```

But document write that `sharedWorkspaceLockfile` Default: true

**Comment by mjbates7** [comment]
I get the same issue but putting what @phuctm97 said above into the .npmrc file got rid of the warning and worked.

Unsure why that is needed as @BabyLy233 said the docs suggest its on by default

**Comment by iltan987** [comment]
Issue still exists. @phuctm97 solution didn't work for me.
I have a simple mono-repo having initial empty Nest.js and Next.js apps. When I try to add a package using "pnpm add", it gives this warning.
Any workaround or solution for this?

**Comment by shahinghasemi** [comment]
Any update on this? I got converted my project into a monorepo workspace, however after each installation I get this annoying warning which is not clear what is all about. After full wiping out node_modules it is not appeared again. Also when the package name is put explicitly in package.json with a manual installation, the warn message does not appear!

**Comment by zkochan** [comment]
This no longer reproduces on current pnpm, so closing.

Verified on pnpm 12.5.1 with `nodeLinker: hoisted`: adding a dependency updates node_modules, not just the lockfile.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
